src/internal/Subscription.ts

```diff
@@ -53,180 +59,201 @@ export class Subscription implements SubscriptionLike {
+        this._teardowns = this._teardowns ?? [];
+        this._teardowns.push(teardown);
       }
     }
+  }
 
-    // Add `this` as parent of `subscription` if that's not already the case.
-    let { _parentOrParents } = subscription;
-    if (_parentOrParents == null) {
-      // If we don't have a parent, then set `subscription._parents` to
-      // the `this`, which is the common case that we optimize for.
-      subscription._parentOrParents = this;
-    } else if (_parentOrParents instanceof Subscription) {
-      if (_parentOrParents === this) {
-        // The `subscription` already has `this` as a parent.
-        return;
-      }
-      // If there's already one parent, but not multiple, allocate an
-      // Array to store the rest of the parent Subscriptions.
-      subscription._parentOrParents = [_parentOrParents, this];
-    } else if (_parentOrParents.indexOf(this) === -1) {
-      // Only add `this` to the _parentOrParents list if it's not already there.
-      _parentOrParents.push(this);
-    } else {
-      // The `subscription` already has `this` as a parent.
-      return;
-    }
+  /**
+   * Checks to see if a this subscription already has a particular parent.
+   * This will signal that this subscription has already been added to the parent in question.
+   * @param parent the parent to check for
+   */
+  private _hasParent(parent: Subscription) {
+    return this._singleParent === parent || this._parents?.includes(parent) || false;
+  }
 
-    // Optimize for the common case when adding the first subscription.
-    const subscriptions = this._subscriptions;
-    if (subscriptions === null) {
-      this._subscriptions = [subscription];
+  /**
+   * Adds a parent to this subscription so it can be removed from the parent if it
+   * unsubscribes on it's own.
+   *
+   * NOTE: THIS ASSUMES THAT {@link _hasParent} HAS ALREADY BEEN CHECKED.
+   * @param parent The parent subscription to add
+   */
+  private _addParent(parent: Subscription) {
+    let { _singleParent, _parents } = this;
+    if (_singleParent) {
+      // We already have one parent so we'll need to expand
+      // to use an array
+      this._parents = [_singleParent, parent];
+      this._singleParent = null;
+    } else if (_parents) {
+      // We already have more than one parent, so just add on to that array.
+      _parents.push(parent);
     } else {
-      subscriptions.push(subscription);
+      // This is our first parent.
+      this._singleParent = parent;
     }
+  }
 
-    return;
+  /**
+   * Called on a child when it is removed via {@link remove}.
+   * @param parent The parent to remove
+   */
+  private _removeParent(parent: Subscription) {
+    if (this._singleParent) {
+      if (this._singleParent === parent) {
+        this._singleParent = null;
+      }
+    } else if (this._parents) {
+      const index = this._parents.indexOf(parent);
+      if (index >= 0) {
+        this._parents.splice(index, 1);
+      }
+    }
```

Comment:
nitpick: There's a fair bit of reading via `this` in here that could be done using destructured variables.